Fix handling consistent targets same as legacy updater

MVrachev opened the pull request:
**Note:** This pr is dependent on pr #1592.

Fixes #1576

**Description of the changes being introduced by the pull request**:

The definition of consistent targets in the spec is ambiguous:
"consistent target files should be written to non-volatile storage
as digest.filename.ext"
Additionally, the specification describes consistent targets when the
client builds the download URL as follows:
"The filename is of the form HASH.FILENAME.EXT".
The issue is about how we interpreted those quotes.
The legacy updater has decided this means a target path `a/b` will translate to a download url path `a/{HASH}.b`.
The ngclient however translates the target path `a/b` to a download URL path `{HASH}.a/b`.

We decided we want to follow the same approach taken from the legacy
updater and thus change how we construct the consistent targets.
Additionally, we want to make sure we test for cases when the TARGETPATH
is an empty string or points to a directory.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Handle consistent targets same as legacy updater

The definition of consistent targets in the spec is ambiguous:
"consistent target files should be written to non-volatile storage
as digest.filename.ext"
Additionally, the specification describes consistent targets when the
client builds the download URL as follows:
"The filename is of the form HASH.FILENAME.EXT".
The issue is about how we interpreted those quotes.
The legacy updater has decided this means a target path "a/b" will
translate to a download url path "a/{HASH}.b".
The ngclient however translates the target path "a/b" to a download url
path "{HASH}.a/b".

We decided we want to follow the same approach taken from the legacy
updater and thus change how we construct the consistent targets.
Additionally, we want to make sure we test for cases when the TARGETPATH
is an empty string or points to a directory.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
While working on this one multiple questions appeared:
1. If a targetpath contains directories, then do we want to create the missing directories? 
There is a separate issue about this and I left a comment there: https://github.com/theupdateframework/python-tuf/issues/1571#issuecomment-926595131
2. How can we create better error messages from the asserts in the for cycles in the unit tests? 
Once one of the asserts failed and I wasn't sure which case failed.
Do we want to use decorators as we did in `test_metadata_serialization.py`?
3. Are there other cases I should cover for TARGETPATH inside the for loops?
4. Do we want to validate the destination directory in download_target?
I think we do need to do that. Are the added checks sufficient? 
5. Don't we want to filter the `target_path` input for `get_one_valid_targetinfo` as well?
6. Are there other file write errors that we should document in the `Raises` for `download_target`
or we can delete one of the TODO items in the Raises section?

jku reviewed (commented):
I think the errors you raise both look like things we could handle earlier: the target path originally comes from the client so if we're validating, let's validate where it really is input -- get_one_valid_targetinfo.

The commit also uses `os.path` to form a URL. It may be possible that Python has enough workarounds to make that work in practice in Windows but I don't like it. `os.path` is for local file paths

jku commented:
> While working on this one multiple questions appeared:
> 
> 1. If a targetpath contains directories, then do we want to create the missing directories?
>    There is a separate issue about this and I left a comment there: [ngclient: downloading targets fails if  targetpath includes subdirectories #1571 (comment)](https://github.com/theupdateframework/python-tuf/issues/1571#issuecomment-926595131)

I'll try to have a PR for that ASAP now that the blocker is merged. But feel free to include the directory creating commit until then

> 2. How can we create better error messages from the asserts in the for cycles in the unit tests?
>    Once one of the asserts failed and I wasn't sure which case failed.
>    Do we want to use decorators as we did in `test_metadata_serialization.py`?

Yeah, that's one of the core advantages of proper table testing, we probably do.

> 4. Do we want to validate the destination directory in download_target?
>    I think we do need to do that. Are the added checks sufficient?

I think we maybe don't need to validate that it exists or can be written to -- I think we can just mention OSError in the docstring. But should test that something reasonable really happens on those cases

I already mentioned the target path checks (if we do them) should IMO happen much earlier.

> 5. Don't we want to filter the `target_path` input for `get_one_valid_targetinfo` as well?

:+1: new issue might be appropriate

MVrachev commented:
> > 5. Don't we want to filter the `target_path` input for `get_one_valid_targetinfo` as well?
> 
> 👍 new issue might be appropriate

Created an issue for that: #1594.

Commit:
Use decorator in test_updater_with_simulator

Reuse the decorator defined in tests/utils.py in order
to receive more helpful messages when an assertion
fails in test_tragets().

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
This pr is ready for review after the merge of #1587 and  #1592.

> > 1. How can we create better error messages from the asserts in the for cycles in the unit tests?
> >    Once one of the asserts failed and I wasn't sure which case failed.
> >    Do we want to use decorators as we did in `test_metadata_serialization.py`?

> Yeah, that's one of the core advantages of proper table testing, we probably do.

I experimented with a new commit where I used the same decorator as `test_metadata_serialization.py`.
IMO it looks good and the error messages are a lot better.

> > 1. Do we want to validate the destination directory in download_target?
> >    I think we do need to do that. Are the added checks sufficient?
> 
> I think we maybe don't need to validate that it exists or can be written to -- I think we can just mention OSError in the docstring. But should test that something reasonable really happens on those cases
> 
> I already mentioned the target path checks (if we do them) should IMO happen much earlier.

I won't focus on validation in this pr.
This should be resolved in #1594.

> The commit also uses os.path to form a URL. It may be possible that Python has enough workarounds to make that work in practice in Windows but I don't like it. os.path is for local file paths

I changed that and instead used a manual file path and URL creation.

@jku can you answer my last question too: 

6. Are there other file write errors that we should document in the Raises for download_target
or we can delete one of the TODO items in the Raises section?

sechkova commented on tests/test_updater_with_simulator.py:
```diff
@@ -82,7 +82,8 @@ def test_refresh(self):
             "targetpath": b"content",
-            "åäö": b"more content"
+            "åäö": b"more content",
+            "dir/targetpath": b"dir target content"
```
Can you add a test case with more that one separator:
"a/b/targetpath"

sechkova commented on tests/test_updater_with_simulator.py:
```diff
@@ -110,8 +111,13 @@ def test_targets(self):
+                # assert local_path != targetpath because of the URL encoding
+                # make target_path absolute as local_path
+                target_path = os.path.join(self.targets_dir, targetpath)
+                self.assertNotEqual(target_path, local_path)
```
I don't know if this test case makes sense, it is too specific and may not be correct when the api changes further. Can we at least move it to a separate test?

MVrachev replied:
I wanted to showcase that the given path is encoded and as a consequence changed if it includes a directory in it.
I decided to use an if statement in order to add as little as possible changes as the code used for the directory case is the same as the other two cases.

Do you still think I should move it to a separate test?

sechkova replied:
Yes, I believe the test cases should be as short and easy to understand as possible and this deserves a separate test.

sechkova commented on tests/test_updater_with_simulator.py:
```diff
@@ -19,6 +19,23 @@
+# Test runner decorator: Runs the test as a set of N SubTests,
+# (where N is number of items in dataset), feeding the actual test
+# function one test case at a time
+def run_sub_tests_with_dataset(dataset: DataSet):
```
Instead of copying the decorator code twice, it would be better to move it to a dedicated tests/util/helper/etc file and reuse it.

MVrachev replied:
I will change my commit description to clarify that the decorator is similar, not the same with the one in `test_metadata_serialization.py`.
First, the `DataSet` type here should be `DataSet = Dict[str, Tuple[str, str]]` instead of `DataSet = Dict[str, str]` which I would fix.
Second, the decorator annotations (including the nested two functions) have different annotations than the ones inside `test_metadata_serialization.py`.

One solution in order to reuse the decorator will be to remove the annotations, but I think annotations are crucial in understanding how the decorator works and I think they should exist and be correct.

jku replied:
This is one of the few places where `Any` annotation makes sense: the dict values are specific to the test and the decorator does not care what the actual type is.

MVrachev replied:
What about the other annotations? I mean the ones we are currently using:
`function: Callable[["TestUpdater", str], None]` and `test_cls: "TestUpdater"`.
Do you think we should use `Any` here too?
As I mentioned above my only concern is if it will be understandable for a new contributor.

jku replied:
There's nothing in those functions that actuall requires a `TestUpdater`, AFAICT. A `unittest.TestCase` should work just fine?

MVrachev replied:
I guess you are right, those classes inherit `unittest.TestCase`, so this can be generalized and still give information with the annotations.

sechkova reviewed (changes requested):
The implementation looks good but I have some reservations for the tests that I put in the comments. Also there isn't really a test for the hash prefix implementation anywhere, the tests rely on the assumption that the default workflow uses consistent targets and no errors were raised.

MVrachev commented:
> Also there isn't really a test for the hash prefix implementation anywhere, the tests rely on the assumption that the default workflow uses consistent targets and no errors were raised.

I will think about how can I enhance the test and leave a comment when I am ready.

jku commented:
> 6\. Are there other file write errors that we should document in the `Raises` for `download_target`
>     or we can delete one of the TODO items in the Raises section?

Sorry I don't understand the question? I don't see any write errors documented currently.

Can file writes fail? Yes. Does our code raise exceptions in those cases? probably: someone needs to read the code and find out

jku commented:
> Also there isn't really a test for the hash prefix implementation anywhere, the tests rely on the assumption that the default workflow uses consistent targets and no errors were raised.

I think this might be fine: we definitely want the default setup to always be consistent_snapshot (test could assert that this is the case, or just add a comment that it's expected).

But I have been thinking that RepositorySimulator could easily have some tracking on what gets requested/downloaded... so a test could then assert that e.g. a specific url was used. This sounds like a potential improvement but maybe a sidetrack for this issue?

Commit:
Tests: move decorator in utils so it can be reused

Generalize the decorator used in test_metadata_serialization.py and
move it inside tests/utils.py, so it can be reused in other similar
situations.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Tests: assert that test_targets use hash prefixes

Make sure that hash prefixes are added when downloading a target
through the repository simulator.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
Updated the pr with the following changes:
- moved the test case for targetpath with directories in a separate function
- moved the decorator in `tests/utils.py` and used it in both `test_metadata_serialization.py` and `test_updater_with_simulator.py`
- added checks that consistent_snapshot` and `prefix_targets_with_hash` are used before downloading a target in the test and a check inside the repo_simulator.

sechkova commented on tests/utils.py:
```diff
@@ -20,8 +20,10 @@
-import argparse
 from contextlib import contextmanager
+from typing import Dict, Any, Callable
+import unittest
```
If the type is quoted ("unittest.TestCase"), this import should not be needed?

MVrachev replied:
It is quoted and it doesn't work without importing `unittest`. 
I think the reason is that quoting is `forward reference`, but it's still a reference that should be resolved: [PEP 0484 forward reference](https://www.python.org/dev/peps/pep-0484/#forward-references).

MVrachev replied:
It is quoted and it doesn't work without importing `unittest`. 
I think the reason is that quoting is `forward reference`, but it's still a reference that should be resolved: [PEP 0484 forward reference](https://www.python.org/dev/peps/pep-0484/#forward-references).

sechkova replied:
Understood. Then probably the quotes can be removed because there will be an unused import which pylint will detect.

MVrachev replied:
I didn't receive unused import, because `unittest` was actually used to resolve the forward reference.
Still, you are correct I should remove the quotes because they are unnecessary.

sechkova replied:
Interesting, I got the error when I applied pylint to the test file.

sechkova commented on tests/repository_simulator.py:
```diff
@@ -184,6 +184,8 @@ def fetch(self, url: str) -> Iterator[bytes]:
             dir_parts, sep , prefixed_filename = path.rpartition("/")
             prefix, _, filename = prefixed_filename.partition(".")
+            if prefix == "":
+                raise ValueError("File name must have hash prefix")
```
Isn't this error duplicating  `NotImplementedError("non-consistent snapshot not supported")`?

jku replied:
I think it's not quite a duplicate, but it's not correct either. Our updaters expect to be able to download non-prefixed targets even if repo is consistent_snapshot: they even have a configuration option to do that... _Maybe_ that was a bad idea considering the unfortunate prefixed-filename-design but it is what Warehouse intends to use.

I don't think a repository can support _completely arbitrary_ target names with this prefix-design (if client asks for target `a.b` we have no way of 100% reliably knowing if `a` is a hash or just part of non-prefixed filename, unless there's some external knowledge of how clients work)... but we just have to implement something in RepositorySimulator so we can test both prefixed and non-prefixed downloads.

Long term maybe there should actually be a `consistent_targets` root variable in the spec to solve this

MVrachev replied:
I don't think so, because if the prefix is "" here this means you have enabled consistent snapshot, but not used hash prefix.
Now when I think for this case this seems to be a possible valid situation - that's why we have the separate `prefix_targets_with_hash` variable, but also the repository simulator doesn't have access to it...
Wondering if I should remove that check considering that the repository simulator doesn't care about the value of `prefix_targets_with_hash` as long as consistent snapshot is True.

MVrachev replied:
> Long term maybe there should actually be a consistent_targets root variable in the spec to solve this

Yes, I like that suggestion. Multiple times I make the mistake to think that `consistent_targets` will be used only if `consistent_snapshot` is True. 
I think it will be beneficial if we split the idea of consistent targets and consistent snapshots.

The question remains: what should we do now? Maybe just remove that check as we don't have enough context to do it well?

jku replied:
yeah the check is definitely not correct

MVrachev replied:
Removed the check and the exception.

sechkova commented on tests/test_updater_with_simulator.py:
```diff
@@ -115,13 +114,40 @@ def test_targets(self, test_case_data: Tuple[str, str]):
-            # make target_path absolute as local_path
-            target_path = os.path.join(self.targets_dir, targetpath)
-            self.assertNotEqual(target_path, local_path)
+    def test_targets_with_subdirectories(self):
```
We've misunderstood each other, sorry about that. Duplicating the whole piece of code like this makes no sense. 
I imagined a couple of lines downloading a file and comparing the paths.

But let's forget about the separate test. I suggest we do something else. In the targets DataSet, the Tuple can be expanded with one more item representing the encoded targethpath. In the ascii case this will duplicate the targetpath. For example:

```python
    targets: utils.DataSet = {
        "standard case": ("targetpath", b"content", "targetpath"),
        "non-asci case": ("åäö", b"more content", "%C3%A5%C3%A4%C3%B6"),
        "subdirectory case": ("a/b/c/targetpath", b"dir target content", "a%2Fb%2Fc%2Ftargetpath"),
    }
```
You can maybe use a named-tuple if you think this is not readable enough. 
At the end of `test_targets`, instead of asserting that the two paths are not equal, we can assert that the local path  really is the encoded string from the DataSet:
`self.assertEqual(local_path, os.path.join(self.targets_dir, encoded_path))`

What do you think about this option?

MVrachev replied:
I like this approach. Will try it.

MVrachev replied:
It worked well I think.
You can have a look now.

sechkova commented on tests/test_updater_with_simulator.py:
```diff
@@ -79,43 +86,45 @@ def test_refresh(self):
+    }
+
+    @utils.run_sub_tests_with_dataset(targets)
+    def test_targets(self, test_case_data: Tuple[str, str]):
```
Since you've defined `class TargetData`, this type annotation is no longer true or at least it should be `Tuple[str, str, str]`. Not sure what would be the opinion of `mypy` actually.

MVrachev replied:
I got no complaints when I was using a `Tuple` and now when I used `TargetData`, but it's better and correct to use `TargetData`.
Thanks for noticing.

sechkova commented on tests/utils.py:
```diff
@@ -49,8 +49,8 @@
 def run_sub_tests_with_dataset(dataset: DataSet):
-    def real_decorator(function: Callable[["unittest.TestCase", Any], None]):
-        def wrapper(test_cls: "unittest.TestCase"):
+    def real_decorator(function: Callable[[unittest.TestCase, Any], None]):
```
The fix should've been in the previous commit but ok.

MVrachev replied:
🤦 will fix that.

sechkova reviewed (approved):
Seems ok now.

jku commented on tuf/ngclient/updater.py:
```diff
@@ -229,6 +229,8 @@ def download_target(
 
         Raises:
+            ValueError: If target_base_url and self._target_base_url are
+                simultaniosly None.
```
typo.

I'd be fine with just saying "ValueError: Invalid arguments" here: the error message will contain the real details and this docstring would likely stay valid if we add other argument validation

jku commented on tests/test_updater_with_simulator.py:
```diff
@@ -79,39 +86,50 @@ def test_refresh(self):
-            "åäö": b"more content"
-        }
+    targets: DataSet = {
+        "standard case": ("targetpath", b"content", "targetpath"),
+        "non-asci case": ("åäö", b"more content", "%C3%A5%C3%A4%C3%B6"),
+        "subdirectory case": ("a/b/c/targetpath", b"dir target content", "a%2Fb%2Fc%2Ftargetpath"),
```
I suppose this works... but you are not actually creating TargetData objects here, just normal tuples

jku reviewed (approved):
Looks fine to me. 

Some observations
* The targets testing is still lacking in volume: we should have vastly more variations on this... but let's do that on it's own issue, as result of #1579 
* I think all of the different DataSet types are _probably_ unneeded now -- it was useful in the very crowded decorator implementation but now the different DataSets might confuse more than help :shrug: 

left two more comments but they are nits as well

jku commented on tests/test_updater_with_simulator.py:
```diff
@@ -19,6 +19,13 @@
 
+class TargetData(NamedTuple):
+    targetpath: str
+    content: str
```
not str btw, it's bytes

MVrachev commented:
> * I think all of the different DataSet types are _probably_ unneeded now -- it was useful in the very crowded decorator implementation but now the different DataSets might confuse more than help 🤷

Yea, maybe you are right, given that as you have pointed out in one of your comments I should create a `TargetData` instance if I want to be correct with the annotations.
Additionally, what is the content of the specific target data becomes clear when I unfold it into specific variables.
I will remove `TargetData` and the specific `DataType` altogether.
